cmd/entrypoint/cmd/init.go

```diff
@@ -0,0 +1,390 @@
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"text/template"
+	"time"
+
+	"github.com/cybozu-go/log"
+	"github.com/cybozu-go/myso"
+	"github.com/cybozu-go/well"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+)
+
+const timeoutSeconds = 30
+
+var (
+	initOnceCompletedPath = filepath.Join(myso.MySQLDataPath, "init-once-completed")
+	socketPath            = filepath.Join(myso.VarRunPath, "mysqld.sock")
+	passwordFilePath      = filepath.Join(myso.TmpPath, "myso-root-password")
+	pingConfPath          = filepath.Join(myso.TmpPath, "ping.cnf")
+)
+
+var initCmd = &cobra.Command{
+	Use:   "init",
+	Short: "Initialize MySQL instance",
+	Long: fmt.Sprintf(`Initialize MySQL instance managed by MySO.
+	If %s already exists, this command does nothing.
+	`, initOnceCompletedPath),
+	SilenceUsage: true,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		well.Go(func(ctx context.Context) error {
+			log.Info("start initialization", nil)
+			err := initializeOnce(ctx)
+			if err != nil {
+				return err
+			}
+
+			log.Info("create config file for ping user", nil)
+			err = confPingUser(ctx)
+			if err != nil {
+				return err
+			}
+
+			log.Info("create config file for admin interface", nil)
+			err = confAdminInterface(ctx, viper.GetString(myso.PodIPFlag))
+			if err != nil {
+				return err
+			}
+
+			log.Info("create config file for server-id", nil)
+			err = confServerID(ctx, viper.GetString(myso.PodNameFlag))
+			if err != nil {
+				return err
+			}
+
+			well.Stop()
+			return nil
+		})
+
+		err := well.Wait()
+		if err != nil && !well.IsSignaled(err) {
+			log.ErrorExit(err)
+		}
+
+		return nil
+	},
+}
+
+func initializeOnce(ctx context.Context) error {
+	_, err := os.Stat(initOnceCompletedPath)
+	if err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	skipInitializeOnce := err == nil
+	if skipInitializeOnce {
+		log.Info("skip data initialization since "+initOnceCompletedPath+" already exists", nil)
+		return nil
+	}
+
+	log.Info("initialize mysql database", nil)
+	err = initializeInstance(ctx)
+	if err != nil {
+		return err
+	}
+
+	log.Info("wait until the instance is started", nil)
+	err = waitInstanceBootstrap(ctx)
+	if err != nil {
+		return err
+	}
+
+	log.Info("setup root user", nil)
+	err = initializeRootUser(ctx, viper.GetString(myso.RootPasswordFlag), viper.GetString(myso.PodIPFlag))
+	if err != nil {
+		return err
+	}
+
+	log.Info("setup operator user", nil)
+	err = initializeOperatorUser(ctx, viper.GetString(myso.OperatorPasswordFlag))
+	if err != nil {
+		return err
+	}
+
+	log.Info("setup operator-admin users", nil)
+	// use the password for an operator-admin user which is the same with the one for operator user
+	err = initializeOperatorAdminUser(ctx, viper.GetString(myso.OperatorPasswordFlag))
+	if err != nil {
+		return err
+	}
+
+	log.Info("setup ping user", nil)
+	err = initializePingUser(ctx)
+	if err != nil {
+		return err
+	}
```

Comment:
This user may mistakenly determine the mysql process is not ready or running
because it does not have any privileges.